File: simply_linked_list/simply_linked_list.c
#include <stdio.h>
#include <stdlib.h>
#include "simply_linked_list.h"

SimplyLinkedList *createSimplyLinkedList()
{
    SimplyLinkedList *list = (SimplyLinkedList *)malloc(sizeof(SimplyLinkedList));

    list->head = NULL;

    return list;
}

void push(SimplyLinkedList *list, int data)
{
    Node *node = createNode(data);

    if (list->head == NULL)
    {
        list->head = node;

        return;
    }

    Node *current = list->head;

    while (getNextNode(current) != NULL)
    {
        current = getNextNode(current);
    }

    setNextNode(current, node);
    return;
}

int pop(SimplyLinkedList *list)
{
    if (list->head == NULL)
        return -1;

    Node *current = list->head;

    int data = current->data;

    list->head = getNextNode(current);

    free(current);

    current = NULL;

    return data;
}

int isEmpty(SimplyLinkedList list)
{
    return list.head == NULL;
}

void printSimplyLinkedList(SimplyLinkedList list)
{
    Node *current = list.head;

    while (current != NULL)
    {
        printf("%d ", current->data);
        current = getNextNode(current);
    }

    printf("\n");
}

void swap(Node *a, Node *b)
{
    int temp = a->data;
    a->data = b->data;
    b->data = temp;
}

void bubbleSort(SimplyLinkedList *list)
{
    Node *current = list->head;

    if (isEmpty(*list))
    {
        return;
    }

    while (current != NULL)
    {
        Node *next = getNextNode(current);

        while (next != NULL)
        {

            if (current->data > next->data)
            {
                swap(current, next);
            }

            next = getNextNode(next);
        }

        current = getNextNode(current);
    }
}

File: structures/simply_linked_list/simply_linked_list.go
package simplylinkedlist

import (
	"errors"
	"fmt"
)

type SimplyLinkedList[T any] struct {
	Head *Node[T]
}

func NewSimplyLinkedList[T any]() *SimplyLinkedList[T] {
	return &SimplyLinkedList[T]{
		Head: nil,
	}
}

func (list *SimplyLinkedList[T]) IsEmpty() bool {
	return list.Head == nil
}

func (list *SimplyLinkedList[T]) Push(value T) {
	node := NewNode(value)

	if list.IsEmpty() {
		list.Head = node
		return
	}

	current := list.Head

	for current.GetNext() != nil {
		current = current.GetNext()
	}

	current.SetNext(node)
}

func (list *SimplyLinkedList[T]) Pop() (T, error) {
	if list.IsEmpty() {
		var v T

		return v, errors.New("empty list")
	}

	current := list.Head

	element := current.GetValue()

	list.Head = current.GetNext()

	return element, nil
}

func swap[T any](a, b *Node[T]) {
	a.Value, b.Value = b.Value, a.Value
}

func (list *SimplyLinkedList[T]) BubbleSort(maximumFunc func(a, b T) bool) {
	current := list.Head

	for list.IsEmpty() {
		return
	}

	for current != nil {
		next := current.GetNext()

		for next != nil {
			if maximumFunc(current.Value, next.Value) {
				swap(current, next)
			}

			next = next.GetNext()
		}

		current = current.GetNext()
	}
}

func (list SimplyLinkedList[T]) String() string {
	if list.Head == nil {
		return "[]"
	}

	current := list.Head
	result := "[ "

	for current != nil {
		result += fmt.Sprintf("%v ", current.Value)
		current = current.GetNext()
	}

	return result + "]"
}